### DISCUSSION TITLE
how to consume spring beans from logback appender?

### DISCUSSION THREAD

**Discussion by sombriks** [qos-ch/logback#687] [Q&A]
Hello all, 

currently i am looking for a a way to offer some sppring beans to a custom appender so i can, for example, log messages to a kafka topic using kafka template.

any guidance on how to perform that?

thanks in advance.

**Comment by ceki** [comment]
@sombriks I have created a [Jira issue]( https://jira.qos.ch/browse/LOGBACK-1757) for tracking the topic of configuration extensions.

**Comment by ceki** [comment]
@sombriks   The question is not specific enough to answer without meandering.

Anyway, appenders can take a custom parameter of any type as long as the parameter type has a default constructor. That should give you a lot of leeway already.

Another rather simple strategy is to chain configurators.  You can configure a custom appender in your custom `ch.qos.logback.core.spi.Configurator` implementation and then delegate to the default configuration system by returning ExecutionStatus.INVOKE_NEXT_IF_ANY from your `configure(Context)` method.

**Reply by sombriks** [reply]
Hi @ceki thanks for put me in the right direction.
I'll see [how spring-boot does that](https://github.com/spring-projects/spring-boot/blob/1a207ed871701a0c889a6b40408702987bf29636/spring-boot-project/spring-boot/src/main/java/org/springframework/boot/logging/logback/SpringBootJoranConfigurator.java) and see if, at some point, is possible to contribute the ApplicationContext itself to it.

**Reply by sombriks** [reply]
Another quick update, Appenders are loaded by a different classloader than the one loading the spring context, right?

i've tried a simple hack using static variables and they kep null, meaning the class i was seeing from spring bean wasn't the same the appender was seeing:

https://github.com/sombriks/sample-kafka/tree/main/service-producer-1/src/main/java/com/example/log2kafka

**Comment by ceki** [comment]
@sombriks It might be more of a initialization order issue rather than a  class loader issue.

**Reply by sombriks** [reply]
i am setting the static attribute yet it remains null, i'll elaborate the test scenario later.

**Reply by maku** [reply]
Do you have a solution for this problem? I tried it but was not able to found one...

**Reply by maku** [reply]
It seems the problem with the class loading comes from the usage of spring-boot-devtools. When I remove this dependency I have no class loading issues anymore and I can the application context in a static variable
